### ISSUE TITLE
Refactoring: Fully support `working_dir` to allow node to reuse chainstate

### ISSUE THREAD

**Issue by zone117x** [blockstack/stacks-blockchain#1576]
Mocknet panics after stopping and running a second time when the `working_dir` is specified.

**Steps To Reproduce**
1. Run 
```BLOCKSTACK_DEBUG=1 RUST_BACKTRACE=full cargo testnet start --config=./mocknet.toml```

   Using `mocknet.toml`:
```toml
[node]
name = "helium-node"
rpc_bind = "0.0.0.0:20443"
working_dir = "/tmp/stacks-blockchain-data"

[burnchain]
chain = "bitcoin"
mode = "mocknet"
commit_anchor_block_within = 2500

[[mstx_balance]]
address = "STB44HYPYAT2BB2QE513NSP81HTMYWBJP02HPGK6"
amount = 10000000000000000

[[mstx_balance]]
address = "ST11NJTTKGVT6D1HY4NJRVQWMQM7TVAR091EJ8P2Y"
amount = 10000000000000000

[[mstx_balance]]
address = "ST1HB1T8WRNBYB0Y3T7WXZS38NKKPTBR3EG9EPJKR"
amount = 10000000000000000

[[mstx_balance]]
address = "STRYYQQ9M8KAF4NS7WNZQYY59X93XEKR31JP64CP"
amount = 10000000000000000
```

2. Let it mine a couple blocks, stop then re-run same command, see panic.
(full logs are attached)
```
INFO [1589368918.400] [testnet/stacks-node/src/tenure.rs:89] Finish tenure: 13e038102e942199629c2441d55748268c04d4093726b6441686bc65c9cbad40
thread 'main' panicked at 'internal error: entered unreachable code', testnet/stacks-node/src/node.rs:564:18
stack backtrace:
DEBUG [1589368918.455] [src/net/p2p.rs:2115] <<<<<<<<<<<<<<<<<<<<<<<<<< End Poll <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
DEBUG [1589368918.456] [src/net/p2p.rs:1640] local.80000000://V4(0.0.0.0:20444): network work state is BlockInvSync
DEBUG [1589368918.456] [src/net/inv.rs:1055] local.80000000://V4(0.0.0.0:20444): Throttle inv sync until 1589368962s
DEBUG [1589368918.456] [src/net/p2p.rs:1640] local.80000000://V4(0.0.0.0:20444): network work state is BlockDownload
DEBUG [1589368918.456] [src/net/p2p.rs:1640] local.80000000://V4(0.0.0.0:20444): network work state is Prune
DEBUG [1589368918.456] [src/net/neighbors.rs:1608] local.80000000://V4(0.0.0.0:20444): walk state is HandshakeBegin
DEBUG [1589368918.456] [src/net/neighbors.rs:1170] local.80000000://V4(0.0.0.0:20444): No neighbors available!  Will not begin neighbor walk
DEBUG [1589368918.456] [src/net/neighbors.rs:1244] Failed to load initial walk neighbors: NoSuchNeighbor
DEBUG [1589368918.568] [src/chainstate/burn/db/burndb.rs:645] Open burndb '/tmp/stacks-blockchain-data/burnchain/db/bitcoin/regtest/burn.db/data.db' as 'readonly', with index as '/tmp/stacks-blockchain-data/burnchain/db/bitcoin/regtest/burn.db/marf'
DEBUG [1589368918.572] [src/net/p2p.rs:2113] >>>>>>>>>>>>>>>>>>>>>>> Begin Poll 1000ms >>>>>>>>>>>>>>>>>>>>>>>>>>>>
   0:        0x10911da05 - std::sys_common::at_exit_imp::push::hee3b7bd54798f075
   1:        0x10913cc6d - core::fmt::ArgumentV1::show_usize::h0555723a314a596c
   2:        0x10911757b - std::io::Write::write_fmt::ha39f6009af02b1d2
   3:        0x10911f81a - std::panicking::default_hook::{{closure}}::h389f076017b5df43
   4:        0x10911f51a - std::panicking::default_hook::h04b06ec20c41bf02
   5:        0x10911fe6d - <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get::h9852a20ed702e745
   6:        0x10915fd7e - serde_json::number::Number::unexpected::h4b16911ae02b2509
   7:        0x1077a2e25 - stacks_node::node::Node::generate_block_commit_op::hda18a8cd5ce86336
                               at testnet/stacks-node/src/node.rs:564
   8:        0x1077a18e4 - stacks_node::node::Node::commit_artifacts::hcd322c990fd1fb49
                               at testnet/stacks-node/src/node.rs:432
   9:        0x10784b4af - stacks_node::run_loop::helium::RunLoop::start::h0029e78efa774c67
                               at testnet/stacks-node/src/run_loop/helium.rs:94
  10:        0x1078cb635 - stacks_node::main::h2c327057e6ee5f90
                               at testnet/stacks-node/src/main.rs:73
  11:        0x107802a02 - std::rt::lang_start::{{closure}}::hac5f13961a864958
                               at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447/src/libstd/rt.rs:67
  12:        0x10911f938 - std::panicking::try::do_call::hf2b35caddf59127c
  13:        0x109123a9b - panic_unwind::dwarf::eh::read_encoded_pointer::h71560c535c128232
  14:        0x1091202ee - std::panicking::rust_panic_without_hook::h0a490548f626bc80
  15:        0x1078029e2 - std::rt::lang_start::h2999b805f1baaed0
                               at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447/src/libstd/rt.rs:67
  16:        0x1078cb9a2 - stacks_node::print_version::ha55c78c4ba1ca12f
```

**Expected behavior**
Ability to resume the mocknet with preserved chain data. 

**Environment (please complete the following information):**
 - Both on Debian and MacOS.
 - Happens using latest master (https://github.com/blockstack/stacks-blockchain/commit/b418a0a2c8aba6181a8139a5e9cc50037145d0a9)
- Happens using master from last week (https://github.com/blockstack/stacks-blockchain/commit/162e49dd1caaf80352108b3cf30c35b9a6d70433)

[first-run.txt](https://github.com/blockstack/stacks-blockchain/files/4621658/first-run.txt)

[second-run.txt](https://github.com/blockstack/stacks-blockchain/files/4621659/second-run.txt)

[stacks-blockchain-data.zip](https://github.com/blockstack/stacks-blockchain/files/4621661/stacks-blockchain-data.zip)

**Comment by diwakergupta** [comment]
@zone117x how critical is this for you now? We haven't made much progress on this PR for a few weeks now and wondering if we can just drop it.

**Comment by diwakergupta** [comment]
Update: we're going to deprioritize this for now.

**Comment by doodlemania2** [comment]
I think this should be addressed as soon as possible so we can iterate more quickly on open critical items without rebuilding from scratch each time we have to restart.

**Comment by sambit4111** [comment]
Yes, agree especially as STX blockchain is aging and growing in size

**Comment by paradigma-cl** [comment]
Some times the fixed working_dir works without problems, but some not.  So in these cases I delete completely the folder and restarts a complete new download and synch. Takes much more time.  When mining the time that takes to recover is very costly.  It should have priority to fix.

**Comment by jcnelson** [comment]
This is addressed in #2370
